### PR TITLE
Avoid optimizing away ref array access

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes https://github.com/dotnet/roslyn/issues/73215.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs:234]: >return arrayAccess.Update(isRef: true); [](http://example.com/codeflow?start=28&length=39)

Is there a good reason to adjust behavior for this scenario? Do we have any debug/release difference in behavior without the change?

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:1010]: >EmitPopIfUnused(used); [](http://example.com/codeflow?start=16&length=22)

I think we cannot leave reference on the stack when a value is expected. Therefore, we should always pop and throw if ```used``` is set

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs:234]: > Do we have any debug/release difference in behavior without the change?

No, there is no debug/release difference neither before nor after the change. Before the change both debug and release emitted `ldelem` or similar; after the change both debug and release emit `ldelema`.

> Is there a good reason to adjust behavior for this scenario?

Using `ref array[x]` and then not using the result (either by not using the local or discarding it explicitly) feels like something that should behave equivalently.

Although now I realized that in case of a ref-readonly assignment we also emit `readonly.ldelema` which doesn't do the runtime element type check, and discards could be considered readonly as well.

So I agree, I will revert this part of the change.

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:1006]: >if (arrayAccess.IsRef) [](http://example.com/codeflow?start=12&length=22)

I think it would be better to add a special node wrapping ```BoundArrayAccess``` rather than adding a new property with unclear effect that could be set improperly. The code in the `if` would go into emit helper for the new node.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:670]: <Field Name="Type" Type="TypeSymbol?" Override="true" Null="always"/>

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:667]: >Needed so we can emit with correct side effects (runtime element type check). [](http://example.com/codeflow?start=4&length=77)

Is the empty line above necessary?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefLocalTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenRefLocalTests.cs:1680]: >IL_0007:  ldelema    ""int"" [](http://example.com/codeflow?start=2&length=28)

Is there a change in behavior for this scenario? I think it is not necessary because when array element is a value type it is not subject for variance and type checks are not necessary.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:1114]: >refArrayAccess.ArrayAccess.Type.IsValueType [](http://example.com/codeflow?start=16&length=43)

Why not put this check to Optimizer and avoid special node creation?

Comment (jjonescz) [reply src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:1114]: Seemed like emit-layer concern about which instruction to emit, but it makes sense to avoid special node creation.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: We are checking `readonly ref` case in Optimizer, but this is also about an instruction choice

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6)

**Comment by jjonescz** [comment]
@dotnet/roslyn-compiler for a second review, thanks

**Review by 333fred — APPROVED** [review]

**Comment by 333fred** [comment]
Rerunning CI since it's been a week, will merge when green.

**Pull request merged**
